src/hotspot/share/runtime/flags/jvmFlagLimit.hpp

```diff
@@ -130,6 +137,29 @@ class JVMFlagLimit {
   }
 
   static JVMFlagConstraintPhase validating_phase() { return _validating_phase; }
+
+  template <typename T>
+  const JVMTypedFlagLimit<T>* cast() const;
+
+  template <typename T>
+  static void assert_compatible_type(int flag_enum) {
+#ifndef PRODUCT
+    if (std::is_integral<T>::value) {
+      switch (flag_enum) {
+      case JVMFlag::TYPE_bool:     assert(sizeof(T) == sizeof(bool)     && std::is_signed<T>::value == std::is_signed<bool>    ::value, "must be"); break;
+      case JVMFlag::TYPE_int:      assert(sizeof(T) == sizeof(int)      && std::is_signed<T>::value == std::is_signed<int>     ::value, "must be"); break;
+      case JVMFlag::TYPE_uint:     assert(sizeof(T) == sizeof(uint)     && std::is_signed<T>::value == std::is_signed<uint>    ::value, "must be"); break;
+      case JVMFlag::TYPE_intx:     assert(sizeof(T) == sizeof(intx)     && std::is_signed<T>::value == std::is_signed<intx>    ::value, "must be"); break;
+      case JVMFlag::TYPE_uintx:    assert(sizeof(T) == sizeof(uintx)    && std::is_signed<T>::value == std::is_signed<uintx>   ::value, "must be"); break;
+      case JVMFlag::TYPE_uint64_t: assert(sizeof(T) == sizeof(uint64_t) && std::is_signed<T>::value == std::is_signed<uint64_t>::value, "must be"); break;
+      case JVMFlag::TYPE_size_t:   assert(sizeof(T) == sizeof(size_t)   && std::is_signed<T>::value == std::is_signed<size_t>  ::value, "must be"); break;
```

Comment:
Could we shorten the signed comparisons to the expected default? For most types the signedness wont be a surprise :) size_t is unsigned on all our platforms. 

is_signed<bool> seems weird, isn't that UB?